### PR TITLE
Show soft kernel crash status

### PR DESCRIPTION
**Pull request by sarab96**
Show soft kernel crash status

### CODE REVIEW

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c [hunk 1]
```diff
@@ -4871,7 +4871,7 @@ kds_custat_show(struct device *dev, struct device_attribute *attr, char *buf)
 		for (;idx < (exec->num_cus + exec->num_sk_cus); ++idx) {
 			sz += sprintf(buf+sz, "CU[@0x0] : %d status : %d name : %s\n",
 				      ert_cu_usage(xert, idx),
-				      ert_cu_status(xert, idx) ? AP_START : AP_IDLE,
+				      ert_cu_status(xert, idx),
```
Comment (larry9523) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c:4874]: Did you test this change on hard CU status? This seems to break the protocol of hard CU status reporting.

Comment (larry9523) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c:4874]: This seems break the AP_IDLE reporting? It used to be '0' representing AP_IDLE (0x4). But now, how are we going to show 0x4 status?

**Review by larry9523 — COMMENTED** [review]
Can we still report correct IDLE status?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c (hunk 1 above)
Comment (sarab96) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c:4874]: It is working fine.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compute Unit Status
CU[  0]: verify:verify_1                 @0x1400000         (IDLE)
SCU[ 0]: hello_world                     @N/A               (--)
SCU[ 1]: hello_world                     @N/A               (--)
SCU[ 2]: hello_world                     @N/A               (CRASHED)
SCU[ 3]: hello_world                     @N/A               (--)
SCU[ 4]: hello_world                     @N/A               (--)
SCU[ 5]: hello_world                     @N/A               (--)
SCU[ 6]: hello_world                     @N/A               (--)
SCU[ 7]: hello_world                     @N/A               (--)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by sarab96** [comment]
Yes. It is working fine.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compute Unit Status
CU[  0]: verify:verify_1                 @0x1400000         (IDLE)
SCU[ 0]: hello_world                     @N/A               (--)
SCU[ 1]: hello_world                     @N/A               (--)
SCU[ 2]: hello_world                     @N/A               (CRASHED)
SCU[ 3]: hello_world                     @N/A               (--)
SCU[ 4]: hello_world                     @N/A               (--)
SCU[ 5]: hello_world                     @N/A               (--)
SCU[ 6]: hello_world                     @N/A               (--)
SCU[ 7]: hello_world                     @N/A               (--)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Can we still report correct IDLE status?

**Comment by larry9523** [comment]
> Yes. It is working fine.
> 
> ```
> Compute Unit Status
> CU[  0]: verify:verify_1                 @0x1400000         (IDLE)
> SCU[ 0]: hello_world                     @N/A               (--)
> SCU[ 1]: hello_world                     @N/A               (--)
> SCU[ 2]: hello_world                     @N/A               (CRASHED)
> SCU[ 3]: hello_world                     @N/A               (--)
> SCU[ 4]: hello_world                     @N/A               (--)
> SCU[ 5]: hello_world                     @N/A               (--)
> SCU[ 6]: hello_world                     @N/A               (--)
> SCU[ 7]: hello_world                     @N/A               (--)
> ```
> 
> > Can we still report correct IDLE status?

Hmm, no, it is not working properly. It used to be

Compute Unit Status
CU[  0]: scaler:scaler_1                 @0x1400000         (IDLE)
CU[  1]: scaler:scaler_2                 @0x1420000         (IDLE)
CU[  2]: lookahead:lookahead_1           @0x1440000         (IDLE)
CU[  3]: lookahead:lookahead_2           @0x1450000         (IDLE)
CU[  4]: decoder:decoder_1               @0x1451000         (IDLE)
CU[  5]: encoder:encoder_1               @0x1452000         (IDLE)
CU[  6]: VCU:VCU_1                       @0x1453000         (IDLE)
SCU[ 0]: kernel1                         @N/A               (IDLE)
SCU[ 1]: kernel1                         @N/A               (IDLE)
SCU[ 2]: kernel1                         @N/A               (IDLE)
SCU[ 3]: kernel1                         @N/A               (IDLE)
SCU[ 4]: kernel1                         @N/A               (IDLE)
SCU[ 5]: kernel1                         @N/A               (IDLE)
SCU[ 6]: kernel1                         @N/A               (IDLE)
SCU[ 7]: kernel1                         @N/A               (IDLE)
SCU[ 8]: kernel1                         @N/A               (IDLE)

**Comment by sarab96** [comment]
This is correct. "--" is correct for zero.
SCU is not showing idle as idle bit is not set (and so showing --).
This is zocl issue (not xbutil). I will submit separate PR for that issue.

> > Yes. It is working fine.
> > ```
> > Compute Unit Status
> > CU[  0]: verify:verify_1                 @0x1400000         (IDLE)
> > SCU[ 0]: hello_world                     @N/A               (--)
> > SCU[ 1]: hello_world                     @N/A               (--)
> > SCU[ 2]: hello_world                     @N/A               (CRASHED)
> > SCU[ 3]: hello_world                     @N/A               (--)
> > SCU[ 4]: hello_world                     @N/A               (--)
> > SCU[ 5]: hello_world                     @N/A               (--)
> > SCU[ 6]: hello_world                     @N/A               (--)
> > SCU[ 7]: hello_world                     @N/A               (--)
> > ```
> > 
> > 
> > > Can we still report correct IDLE status?
> 
> Hmm, no, it is not working properly. It used to be
> 
> Compute Unit Status
> CU[ 0]: scaler:scaler_1 @0x1400000 (IDLE)
> CU[ 1]: scaler:scaler_2 @0x1420000 (IDLE)
> CU[ 2]: lookahead:lookahead_1 @0x1440000 (IDLE)
> CU[ 3]: lookahead:lookahead_2 @0x1450000 (IDLE)
> CU[ 4]: decoder:decoder_1 @0x1451000 (IDLE)
> CU[ 5]: encoder:encoder_1 @0x1452000 (IDLE)
> CU[ 6]: VCU:VCU_1 @0x1453000 (IDLE)
> SCU[ 0]: kernel1 @N/A (IDLE)
> SCU[ 1]: kernel1 @N/A (IDLE)
> SCU[ 2]: kernel1 @N/A (IDLE)
> SCU[ 3]: kernel1 @N/A (IDLE)
> SCU[ 4]: kernel1 @N/A (IDLE)
> SCU[ 5]: kernel1 @N/A (IDLE)
> SCU[ 6]: kernel1 @N/A (IDLE)
> SCU[ 7]: kernel1 @N/A (IDLE)
> SCU[ 8]: kernel1 @N/A (IDLE)

**Comment by sarab96** [comment]
Update to show zero soft kernel status as idle. Zocl changes to set idle bit for soft kernels can be handled separately

File: src/runtime_src/core/pcie/tools/xbutil/xbutil.h [hunk 2]
```diff
@@ -463,10 +463,13 @@ class device {
             sensor_tree::add_child( std::string("board.compute_unit." + std::to_string(i)), ptCu );
         }
 
+        //Soft kernel info below
         for (unsigned int i = computeUnits.size(); i < parseComputeUnitNum(custat); i++) {
             uint32_t status = parseComputeUnitStat(custat, i, cu_stat::stat);
             uint32_t usage = parseComputeUnitStat(custat, i, cu_stat::usage);
 	    auto name = parseComputeUnitName(custat, i);
+            if (!status)
+                status = 0x4;//zero is idle for soft kernels; zocl only send 0,1,-1 status
```
Comment (larry9523) [inline src/runtime_src/core/pcie/tools/xbutil/xbutil.h:472]: I think we'd better make the meaning of sysfs node consistent. By saying that, not just xbutil, any other host code, sysfs node consumer can have a consistent way to interpret cu/scu status.
I suggest keep 0x4 in sysfs node for AP_IDEL status.

File: src/runtime_src/core/common/utils.cpp [hunk 3]
```diff
@@ -43,37 +43,41 @@ parse_cu_status(unsigned int val)
...
-    delim = '|';
-  }
-  if (val & 0x10) {
-    status += delim;
-    status += "RESTART";
-    delim = '|';
-  }
-  if (status.size())
-    status += ')';
+  if (val == (uint32_t)-1) //Crashed soft kernel status is -1
```
Comment (stsoe) [inline src/runtime_src/core/common/utils.cpp:46]: No c-style casting.  Use std::numeric_limits<uint32_t>::max();

File: src/runtime_src/core/common/utils.cpp [hunk 4]
```diff
@@ -43,37 +43,41 @@ parse_cu_status(unsigned int val)
...
+    if (val & 0x10) {
+      status += delim;
+      status += "RESTART";
+      delim = '|';
+    }
+    if (status.size())
+      status += ')';
+    else 
+      status = "(UNKNOWN)";
+  }
```
Comment (stsoe) [inline src/runtime_src/core/common/utils.cpp:80]: It'd be better if we some how defined a symbolic name for all these 0xNN constants, e.g AP_START, AP_DONE, etc.

File: src/runtime_src/core/pcie/tools/xbutil/xbutil.h (hunk 2 above)
Comment (stsoe) [reply src/runtime_src/core/pcie/tools/xbutil/xbutil.h:472]: Can we please consolidate AP_IDLE, AP_START, etc.  0x4 is just weird looking.

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c (hunk 1 above)
Comment (larry9523) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c:4874]:
```suggestion
				      !ert_cu_status(xert, idx) ? AP_IDLE : (ert_cu_status == 1 ? AP_START : -1),```

**Review by larry9523 — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/pcie/tools/xbutil/xbutil.h (hunk 2 above)
Comment (sarab96) [reply src/runtime_src/core/pcie/tools/xbutil/xbutil.h]: done

File: src/runtime_src/core/common/utils.cpp (hunk 3 above)
Comment (sarab96) [reply src/runtime_src/core/common/utils.cpp]: done

File: src/runtime_src/core/common/utils.cpp (hunk 4 above)
Comment (sarab96) [reply src/runtime_src/core/common/utils.cpp]: done

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c (hunk 1 above)
Comment (sarab96) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/mb_scheduler.c]: done

**Review by larry9523 — APPROVED** [review]

**Review by stsoe — APPROVED** [review]

**Review by mamin506 — APPROVED** [review]

**Pull request merged**